### DISCUSSION TITLE
pkgutil.iter_modules extremely slow in 6.1.0

### DISCUSSION THREAD

**Discussion by markgemmill** [pyinstaller/pyinstaller#8058] [Help]
After upgrading to 6.1.0 pkgutil.iter_modules has become extremely slow. At startup my app recursively scans all the modules to make sure certain ones are imported. When running unfrozen, or frozen with 5.13, this takes less than a second to execute.  Under 6.10 it now takes a couple of minutes. 

Any ideas on what might be causing this with 6.1.0?

**Comment by rokm** [comment]
By 5.13, do you mean 5.13.0, or also 5.13.1 and 5.13.2?

Because at a quick glance, I would say https://github.com/pyinstaller/pyinstaller/commit/afc55a868cfafb427b50c4379a889c078c58951d might be a candidate, but we back-ported that to 5.13.1 as well...

Can you provide an example with `pkgutil.iter_modules` as you use it, so I can test locally?

**Reply by markgemmill** [reply]
After digging into the change history, the last version of pyinstaller I actually ran and tested under was likely 4.10.   

Here's the class I'm using that's making the pkgutil.iter_modules call. 

```python
import inspect
import os
import pkgutil
import sys
from importlib import import_module
from pathlib import Path


class ModuleManager:
    """
    usage:

        packages = ModuleManager2(apiserver, 'task', LazyActor)
        packages.find_task_packages()

    """

    def __init__(self, module, pkg_name, task_type):
        self.frozen = getattr(sys, "frozen", False)
        self.root_path = Path(module.__file__).parent
        self.pkg_name = pkg_name
        self.task_type = task_type
        if self.frozen:
            self.FRZ_PREFIX_LEN = len(str(sys._MEIPASS) + os.path.sep)

    @property
    def app_name(self):
        if self.frozen:
            return str(self.root_path)
        return self.root_path.name

    def get_module_name(self, pkg_path):
        if self.frozen:
            pkg_path = pkg_path[self.FRZ_PREFIX_LEN :]
        return pkg_path.replace(os.path.sep, ".")

    def find_task_packages(self, package_name: str = None, load=False):
        """
        Iterate over the application package and look
        for any package or module named task and import it.

        Print out the name of all dramatiq Actor functions.

        """
        if package_name is None:
            package_name = self.app_name

        # FIX: slow-down starts here....
        for module_info in pkgutil.iter_modules([package_name]):

            # FIX: slowdown ends here... so something with pkgutil.iter_modules is super slow

            module_path = f"{package_name}{os.path.sep}{module_info.name}"

            # we're looking for just 'task' packages
            do_load = module_info.name == "tasks" or load is True

            if not module_info.ispkg and do_load:
                module_name = self.get_module_name(module_path)
                _module = import_module(module_name)

                # COMMENT: this bit is just informational
                for name, obj in inspect.getmembers(_module):
                    if isinstance(obj, self.task_type):
                        print(f"[task object] {module_name}.{name}")
                continue

            self.find_task_packages(module_path, load=do_load)
```

**Reply by rokm** [reply]
> After digging into the change history, the last version of pyinstaller I actually ran and tested under was likely 4.10.

Hmm, can you test with 5.13.0, then? Is that one slow already? If not, what about 5.13.1?

**Reply by markgemmill** [reply]
I'm looking at py_rth_pkgutil.py and will dive into what's going on there and see if I can't figure out what's happening.

**Reply by rokm** [reply]
> I'm looking at py_rth_pkgutil.py and will dive into what's going on there and see if I can't figure out what's happening.

OK, thanks!

**Reply by markgemmill** [reply]
I'll give 5.13+ a go as well.

**Reply by markgemmill** [reply]
I see what's going on. I'm traversing my application's packages by recursively calling `pkgutil.iter_modules`. This works fine with the stdlib call, but the `_pyi_pkgutil_iter_modules `version ends up looping through the entire `PyiFrozenImporter.toc` for each call, which in my case is 2500+ toc entries * 175 packages.

**Reply by rokm** [reply]
> I see what's going on. I'm traversing my application's packages by recursively calling `pkgutil.iter_modules`. This works fine with the stdlib call, but the `_pyi_pkgutil_iter_modules `version ends up looping through the entire `PyiFrozenImporter.toc` for each call, which in my case is 2500+ toc entries * 175 packages.

Hmm, but in that case, this must have been an issue ever since the run-time hook was added (which I believe was in PyInstaller 4.4), since it has been iterating over the entire TOC since the very first version?

Or has the performance regressed in some subsequent version? And if so, in which?

I'm asking because if this is a regression, we could take a look at offending commit and try to remedy it. But if it has been an issue from the beginning, we will probably have to rethink the approach (for example, move the trie implementation from `pyi_rth_pkgres` into `PyiFrozenImporter`, and have `pyi_rth_pkgutil` use it to iterate only over relevant part of the TOC).

**Reply by markgemmill** [reply]
The difference between 4.10 and 6.10 appears to be this added line:

```python
module_path = os.path.realpath(os.path.join(MEIPASS, entry.replace(".", os.path.sep)))
```

Otherwise, the code inside the `toc` loop is pretty much the same.  In my use-case that line executing 400,000+ times.  I ran a simulation and that pretty much accounts for the lag I'm seeing.

If this has been there since 5.13.1 and I'm the only one that is experiencing an issue so far, is it worth trying to improve the `iter_modules` hook at this point in time?  

I have an idea about how I'll work around this for my use-case, so I think I'll be fine. I'll also post what I come up with and maybe that will give some ideas for a future improvement.


6.1.0
```python
                for entry in importer.toc:
                    # Get the path to the module, and resolve symbolic links. This implicitly solves the problem with
                    # macOS .app bundles and packages that contain data files, where package directory is fully
                    # collected into Contents/Resources, and symbolic link to directory is created in sys._MEIPASS
                    # (Contents/Frameworks). See #7884.

                    module_path = os.path.realpath(os.path.join(MEIPASS, entry.replace(".", os.path.sep)))

                    # If the module is not in the requested path, skip it.
                    if not module_path.startswith(pkg_path):
                        continue

                    name = module_path[len(pkg_path):]

                    if os.path.sep in name:
                        # Not a direct child
                        continue
                    is_pkg = importer.is_package(entry)

                    yield pkgutil.ModuleInfo(importer, prefix + name, is_pkg)
```

4.10
```python
            for entry in importer.toc:
                if not entry.startswith(pkg_prefix):
                    continue
                name = entry[pkg_prefix_len:]
                if name.count('.') != 0:
                    continue
                is_pkg = importer.is_package(entry)
                yield pkgutil.ModuleInfo(importer, prefix + name, is_pkg)
```

**Reply by rokm** [reply]
Thanks for the analysis! 

This means that we should go back to checks being done in the module-name space rather than path-space, because then the `os.path.realpath` resolution needs to be done only once (or twice, for macOS .app bundles) per given search path, instead of once per toc entry. (And we could still also look into applying trie built on top of the toc entries, which I expect would amortize across multiple `iter_modules` calls). 

> If this has been there since 5.13.1 and I'm the only one that is experiencing an issue so far, is it worth trying to improve the iter_modules hook at this point in time?

Yeah, it's not a exactly a high-priority issue, but it would still be nice to sort it out, now that we know what the cause is.

**Comment by markgemmill** [comment]
@rokm Thanks for pointing me in the right direction, not to mention maintaining this project. 

I'll post back what I come up with and, with your guidance, maybe I can help out with a more permanent solution.

**Reply by rokm** [reply]
Here's a branch with comparison moved back into module-name space: https://github.com/rokm/pyinstaller/tree/pkgutil-iter_module-performance

Can you check if it improves execution time, and if it is comparable to 4.10?

``` 
pip uninstall pyinstaller
pip install git+https://github.com/rokm/pyinstaller@pkgutil-iter_module-performance
```
or 
``` 
pip uninstall pyinstaller
pip install https://github.com/rokm/pyinstaller/archive/refs/heads/pkgutil-iter_module-performance.zip
```

**Reply by markgemmill** [reply]
Downloaded your branch, installed and re-built, but I'm get this error when I run the exe:
This is a windows environment.

```python
Traceback (most recent call last):
  File "apiserver_exe.py", line 2, in <module>
  File "PyInstaller\loader\pyimod02_importers.py", line 391, in exec_module
  File "apiserver\cli.py", line 3, in <module>
    from .api import create_app
  File "PyInstaller\loader\pyimod02_importers.py", line 391, in exec_module
  File "apiserver\api.py", line 17, in <module>
    from . import unfi
  File "PyInstaller\loader\pyimod02_importers.py", line 391, in exec_module
  File "apiserver\unfi\__init__.py", line 9, in <module>
    from . import purchasing
  File "PyInstaller\loader\pyimod02_importers.py", line 391, in exec_module
  File "apiserver\unfi\purchasing\__init__.py", line 1, in <module>
    from .routes import api  # noqa
    ^^^^^^^^^^^^^^^^^^^^^^^
  File "PyInstaller\loader\pyimod02_importers.py", line 391, in exec_module
  File "apiserver\unfi\purchasing\routes.py", line 11, in <module>
    from . import tasks
  File "PyInstaller\loader\pyimod02_importers.py", line 391, in exec_module
  File "apiserver\unfi\purchasing\tasks.py", line 8, in <module>
    from mrsx.api import fetch_report as fetch_mrs_report
  File "PyInstaller\loader\pyimod02_importers.py", line 391, in exec_module
  File "mrsx\api.py", line 6, in <module>
  File "PyInstaller\loader\pyimod02_importers.py", line 391, in exec_module
  File "requests_ntlm\__init__.py", line 1, in <module>
  File "PyInstaller\loader\pyimod02_importers.py", line 391, in exec_module
  File "requests_ntlm\requests_ntlm.py", line 11, in <module>
  File "PyInstaller\loader\pyimod02_importers.py", line 391, in exec_module
  File "spnego\__init__.py", line 26, in <module>
  File "PyInstaller\loader\pyimod02_importers.py", line 391, in exec_module
  File "spnego\auth.py", line 10, in <module>
  File "PyInstaller\loader\pyimod02_importers.py", line 391, in exec_module
  File "spnego\_negotiate.py", line 31, in <module>
  File "PyInstaller\loader\pyimod02_importers.py", line 391, in exec_module
  File "spnego\_sspi.py", line 39, in <module>
  File "PyInstaller\loader\pyimod02_importers.py", line 391, in exec_module
  File "sspilib\__init__.py", line 6, in <module>
  File "PyInstaller\loader\pyimod02_importers.py", line 391, in exec_module
  File "sspilib\raw\__init__.py", line 8, in <module>
  File "src\\sspilib\\raw\\_context_attributes.pyx", line 1, in init sspilib.raw._context_attributes
  File "src\\sspilib\\raw\\_security_context.pyx", line 1, in init sspilib.raw._security_context
  File "src\\sspilib\\raw\\_credential.pyx", line 1, in init sspilib.raw._credential
ModuleNotFoundError: No module named 'sspilib.raw._text'
[21656] Failed to execute script 'apiserver_exe' due to unhandled exception!
```

**Reply by rokm** [reply]
That's not related to the branch - the `sspilib.raw` has multiple cythonized extensions, so some of its modules are not picked up automatically. Add the missing module (`sspilib.raw._text`) to hidden imports. Or perhaps collect all submodules (`--collect-submodules sspilib.raw`) if multiple modules from that package are missing.

**Reply by markgemmill** [reply]
That did the trick, and the exe built with that branch is back to regular speed. Might even be a bit quicker. :-)
